cuda/context_manager.cpp

```diff
@@ -0,0 +1,258 @@
+  return true;
+}
+
+bool CUDAContextManager::get_stream_device(hpt_t hpt, CUstream stream, CUdevice *pdev) {
+  if (stream == 0) {
+    // using the default stream in the top context on the stack
+    return get_device(hpt, pdev);
+  }
+  hp_stream_t hp_stream_key = {std::get<0>(hpt), std::get<1>(hpt), stream};
+  CUcontext ctx;
+  ctx = THAPI_AT(hp_stream_to_ctx_, hp_stream_key);
+  auto ctx_find = hp_ctx_to_device_.find({std::get<0>(hpt), std::get<1>(hpt), ctx});
+  if (ctx_find != hp_ctx_to_device_.end()) {
+    *pdev = ctx_find->second;
+    return true;
+  }
+  return false;
+}
+
+// cuPrimaryCtxRetain_v2_(entry|exit)
+void CUDAContextManager::primary_ctx_retain_entry(hpt_t hpt, CUdevice dev) {
+  // std::cerr << "primaryctx retain entry thread " << std::get<2>(hpt) << std::endl;
+  hpt_entry_dev_[hpt] = dev;
+}
+
+void CUDAContextManager::primary_ctx_retain_exit(hpt_t hpt, CUresult cuResult,
+                                                 CUcontext ctx) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  // std::cerr << "primaryctx retain exit thread " << std::get<2>(hpt) << std::endl;
+  try {
+    auto entry_dev = THAPI_AT(hpt_entry_dev_, hpt);
+    hp_device_primary_ctx_[{std::get<0>(hpt), std::get<1>(hpt), entry_dev}] = ctx;
+    hp_ctx_to_device_[{std::get<0>(hpt), std::get<1>(hpt), ctx}] = entry_dev;
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no entry device in primary ctx retain exit callback");
+  }
+}
+
+
+// cuPrimaryCtxRelease_v2_(entry|exit)
+void CUDAContextManager::primary_ctx_release_entry(hpt_t hpt, CUdevice dev) {
+  hpt_entry_dev_[hpt] = dev;
+}
+
+void CUDAContextManager::primary_ctx_release_exit(hpt_t hpt, CUresult cuResult) {
+  if (cuResultIsError(cuResult)) {
+    return;
+  }
+  CUdevice entry_dev;
+  try {
+    entry_dev = THAPI_AT(hpt_entry_dev_, hpt);
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no entry device in primary ctx release exit callback");
+  }
+  // std::cerr << "release primary ctx for device " << entry_dev << " thread " << std::get<2>(hpt) << std::endl;
+  hp_device_t hp_dev_key = {std::get<0>(hpt), std::get<1>(hpt), entry_dev};
+  try {
+    auto ctx = THAPI_AT(hp_device_primary_ctx_, hp_dev_key);
+    hp_device_primary_ctx_.erase(hp_dev_key);
+    hp_ctx_to_device_.erase({std::get<0>(hpt), std::get<1>(hpt), ctx});
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+                    "no context found in primary ctx release exit callback");
+  }
+}
+
+
+// cuPrimaryCtxReset_v2_(entry|exit)
+void CUDAContextManager::primary_ctx_reset_entry(hpt_t hpt, CUdevice dev) {
+  primary_ctx_release_entry(hpt, dev);
+}
+
+void CUDAContextManager::primary_ctx_reset_exit(hpt_t hpt, CUresult cuResult) {
+  primary_ctx_release_exit(hpt, cuResult);
+}
+
+
+// cuCtxCreate_v(2|3)_(entry|exit)
+void CUDAContextManager::ctx_create_entry(hpt_t hpt, CUdevice dev) {
+  hpt_entry_dev_[hpt] = dev;
```

Comment:
Should we use an internal the newly introduced `EntryState`  to do so?

Reply:
Good idea, I will see if I can easily pass it to the constructor or maybe set it after the fact in the filter init.